cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/CSeq.scala

```diff
@@ -0,0 +1,158 @@
+package com.twitter.cassovary.util.collections
+
+import scala.collection.immutable
+
+trait CSeq[@specialized(Int, Long) T] {
+  def apply(idx: Int): T
+  def length: Int
+
+  def slice(from: Int, until: Int)(implicit csf: CSeqFactory[T]) : CSeq[T] = {
+    val new_until = math.min(length, until)
+    if (from >= new_until) {
+      CSeq.empty[T]
+    } else {
+      CSeq.apply[T](this, from, new_until)
+    }
+  }
+
+  def toSeq: CSeqWrapper[T] = new CSeqWrapper[T](this)
+
+  def isEmpty = length == 0
+
+  def foreach[U](f: T => U): Unit = {
+    var i: Int = 0
+    while (i < length) {
+      f(apply(i))
+      i = i + 1
+    }
+  }
+
+  override def hashCode(): Int = toSeq.hashCode()
+
+  override def equals(obj: scala.Any): Boolean = obj match {
+    case cseq: CSeq[T] => toSeq.equals(cseq.toSeq)
+    case _ => false
+  }
+}
+
+class CSeqWrapper[T](underlying: CSeq[T]) extends immutable.IndexedSeq[T] {
+  override def length: Int = underlying.length
+
+  override def apply(idx: Int): T = underlying.apply(idx)
+}
+
+object CSeq {
+  def ofDim[@specialized(Int, Long) T](n: Int)(implicit cSeqFactory: CSeqFactory[T]) = {
+    cSeqFactory.ofDim(n)
+  }
+
+  def apply[@specialized(Int, Long) T](arr: Array[T])(implicit cSeqFactory: CSeqFactory[T]) = {
+    cSeqFactory.fromArray(arr)
+  }
+
+  def apply[@specialized(Int, Long) T](arr: Array[T], from: Int, until: Int)
+                                      (implicit cSeqFactory: CSeqFactory[T]): CSeq[T] = {
+    cSeqFactory.fromCSeq(apply(arr), from, until)
+  }
+
+
+  def apply[@specialized(Int, Long) T](arr: CSeq[T], from: Int, until: Int)
+                                      (implicit cSeqFactory: CSeqFactory[T]): CSeq[T] = {
+    cSeqFactory.fromCSeq(arr, from, until)
+  }
+
+  def empty[@specialized(Int, Long) T](implicit cSeqFactory: CSeqFactory[T]) = {
+    cSeqFactory.empty
+  }
+
+  def apply(x: Int, xs: Int*): CSeq[Int] = {
+    import Implicits._
+    CSeq[Int](Array(x, xs:_*))
+  }
+
+  class IntAndLongImplicits {
+    val cSeqFactoryInt: CSeqFactory[Int] = new CSeqFactoryInt
+    implicit def csi: CSeqFactory[Int] = cSeqFactoryInt
+    implicit val cSeqFactoryLong: CSeqFactory[Long] = new CSeqFactoryLong
+  }
+
+  object Implicits extends IntAndLongImplicits
+}
+
+abstract class CSeqFactory[@specialized(Int, Long) T] {
```

Comment:
When using Macros this wasn't that easy. I needed CSeq, CSeqFactory and CSeqWrapper in the same file to be visible by the macro.